### PR TITLE
Add new logging API for events with custom fields

### LINKED ISSUE #1654: Extend logger interface with fields logging
Currently, we only have log single string messages, e.g.,:

```
caf::logger::debug("app", "Hello, {}!", "World");
```

If we want to support more advanced logging schemes (like ECS), we'll also need an API to log multiple fields per event, e.g.:

```
caf::logger::debug_fields("app", {{"event.code", "200"}, {"event.id", "82de72"}});
caf::logger::debug_fields_from("app", my_object); // uses `inspect` to get fields from `my_object`
```

### PR DESCRIPTION
**Pull request by Neverlord**
Closes #1654.

### CODE REVIEW

**Review by riemass — CHANGES_REQUESTED** [review]
My only review item is that some `field` member functions aren't covered in tests. Otherwise looks good and can be merged. 

What I don't like is that we have the cost of calling 'logger::trace' while running on a less verbose logger. Things like `logger::accepts` are virtual function calls. What do you think of benchmarking the macro based API vs the new API for ~5 million log messages?

There are a few loose ends. I think we should update the manual with the new logger API before the release. Also when we integrate this, we should remove the macro logs from our  side to lead with a good example for the users. I grepped the `examples` folder for the word `log` and it seems like we don't have any user logging examples in place. 

And finally a feature request for structured logging, stuff like:
```
auto response = my_future.get();
logger::debug("app", "Received response.", response);
```
It should be possible to make this with our inspect API and map the struct to fields lazily.
Solutions like 
```
logger::debug("app", "Received response.", to_string(response));
```
should be discouraged because we pay the price for stratification at call site, even when we don't log this verbosity. 

The last couple of PRs are great progress and I was glad to review them. Looking forward for version v1.0. :steam_locomotive: :train: :train:.

File: libcaf_core/caf/log_event.cpp [hunk 1]
```diff
@@ -22,13 +23,70 @@ std::string_view deep_copy_impl(detail::monotonic_buffer_resource* resource,
...
+  return chunked_string{head};
+}
+
 } // namespace
 
+log_event_ptr log_event::with_message(std::string_view msg,
+                                      keep_timestamp_t) const {
+  // Note: can't use make_counted here because the constructor is private.
+  auto copy = log_event_ptr{new log_event, false};
+  auto* resource = &copy->resource_;
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/log_event.cpp:58]: No, the resource is per-event.

**Comment by Neverlord** [comment]
> My only review item is that some `field` member functions aren't covered in tests. Otherwise looks good and can be merged.

Pushes a new test that should cover the missing blocks.

> What I don't like is that we have the cost of calling 'logger::trace' while running on a less verbose logger. Things like `logger::accepts` are virtual function calls. What do you think of benchmarking the macro based API vs the new API for ~5 million log messages?

Tracing is definitely expensive, that's why I'd rather not use `logger::trace` directly (cf. #1652):

> The only logging that I would still keep behind a CMake build option is trace logging. Trace logs are expensive and traces into internal CAF functions are (with some very rare exceptions) only useful when working on CAF itself. I think it's perfectly fine to require custom builds for this use case.

That being said, benchmarking is always a good idea. But we'd have to compare to the 0.19.4 implementation.

> There are a few loose ends. I think we should update the manual with the new logger API before the release. Also when we integrate this, we should remove the macro logs from our side to lead with a good example for the users. I grepped the `examples` folder for the word `log` and it seems like we don't have any user logging examples in place.

Yes, switching to the new functions is definitely a next step. We can also think about deprecating the macros with 1.0.

> And finally a feature request for structured logging, stuff like:
> 
> ```
> auto response = my_future.get();
> logger::debug("app", "Received response.", response);
> ```
> 
> It should be possible to make this with our inspect API and map the struct to fields lazily.

Full ack. The `detail::format` indirection can probably wrap types that are inspectable but don't specialize `std::formatter`. For this, we'll probably have to re-implement `detail::stringification_inspector` to work on output iterators. Otherwise, we'll also produce an intermediary string.

File: libcaf_core/caf/log_event.cpp [hunk 2]
```diff
@@ -22,13 +23,70 @@ std::string_view deep_copy_impl(detail::monotonic_buffer_resource* resource,
...
+chunked_string::node_type*
+deep_copy_to_node(detail::monotonic_buffer_resource* resource,
+                  std::string_view str) {
+  using node_type = chunked_string::node_type;
+  auto* buf = allocator_t<node_type>{resource}.allocate(1);
+  return new (buf) node_type{deep_copy_impl(resource, str), nullptr};
+}
+
+chunked_string deep_copy_impl(detail::monotonic_buffer_resource* resource,
+                              chunked_string str) {
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_core/caf/log_event.cpp:35]: Done. Thanks for keeping me honest on the tests. 🙂

**Review by riemass — APPROVED** [review]
This looks good now. Full ack on deprecating log macros with 1.0.

**Pull request merged**